src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SpeculativeTCompletionProviderTests.cs

```diff
@@ -162,6 +162,454 @@ class C
+        {
+            var markup = @"
+using System;
+class C
+{
+    private ref Func<System.Func<int,$$
+}";
+
+            await VerifyItemExistsAsync(markup, "T");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.Completion)]
+        [WorkItem(37224, "https://github.com/dotnet/roslyn/issues/37224")]
+        [WorkItem(37268, "https://github.com/dotnet/roslyn/issues/37268")]
+        public async Task InRefAndQualifiedNestedGeneric2()
+        {
+            var markup = @"
+using System;
+class C
+{
+    public ref Func<int, System.Func<int,$$
+}";
+
+            await VerifyItemExistsAsync(markup, "T");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.Completion)]
+        [WorkItem(37224, "https://github.com/dotnet/roslyn/issues/37224")]
+        [WorkItem(37268, "https://github.com/dotnet/roslyn/issues/37268")]
+        public async Task InRefAndQualifiedNestedGeneric3()
+        {
+            var markup = @"
+using System;
+class C
+{
+    private protected ref Func<int, System.Func<$$
+}";
+
+            await VerifyItemExistsAsync(markup, "T");
+        }
+
+        [Fact(Skip = "https://github.com/dotnet/roslyn/issues/37361"), Trait(Traits.Feature, Traits.Features.Completion)]
+        public async Task InTuple0()
+        {
+            var markup = @"
+using System;
+class C
+{
+    protected ($$
+}";
+
+            await VerifyItemExistsAsync(markup, "T");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.Completion)]
+        public async Task InTupleNot0()
+        {
+            var markup = @"
+using System;
+class C
+{
+    protected sealed (int $$
+}";
+
+            await VerifyItemIsAbsentAsync(markup, "T");
+        }
+
+        [Fact(Skip = "https://github.com/dotnet/roslyn/issues/37361"), Trait(Traits.Feature, Traits.Features.Completion)]
+        public async Task InTuple1()
+        {
+            var markup = @"
+using System;
+class C
+{
+    sealed (int, $$
+}";
+
+            await VerifyItemExistsAsync(markup, "T");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.Completion)]
+        public async Task InTupleNot1()
+        {
+            var markup = @"
+using System;
+class C
+{
+    virtual (int x, C $$
+}";
+
+            await VerifyItemIsAbsentAsync(markup, "T");
+        }
+
+        [Fact(Skip = "https://github.com/dotnet/roslyn/issues/37361"), Trait(Traits.Feature, Traits.Features.Completion)]
+        public async Task InTupleGeneric0()
+        {
+            var markup = @"
+using System;
+class C
+{
+    (Func<$$
+}";
+
+            await VerifyItemExistsAsync(markup, "T");
+        }
+
+        [Fact(Skip = "https://github.com/dotnet/roslyn/issues/37361"), Trait(Traits.Feature, Traits.Features.Completion)]
+        public async Task InTupleGeneric1()
+        {
+            var markup = @"
+using System;
+class C
+{
+    (int, Func<$$
+}";
+
+            await VerifyItemExistsAsync(markup, "T");
```

Comment:
more preferable would be to mark it with WorkItem, have it be non skipped, test the *current* behavior, but put comment that this is broken.

Reply:
I don't understand why we write tests to 'expect' the unexpected behavior preferable. Making them skipped at least give us warning from Test Explorer showing that we have more work to do, instead of all green. For example, there are several cases skipped for local function context at the bottom of this file

https://github.com/dotnet/roslyn/blob/759e0b69ecc2ae5a700be05c6d27dfaa8c43a3e4/src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SpeculativeTCompletionProviderTests.cs#L360-L373)

I'm ok with marking them with WorkItem.

Reply:
Primary reasons:

1. skipped tests almost never get unskipped.  so they're effectively useless.
2. a test is simply a way of recording the expected behavior.  right now, the behavior is expected that we don't get `T` since we have a bug.  Once we fix that, this test will actually fire, demonstrating that yes, we did actually do what we set out to do.

Reply:
I don't think we call the recording 'expected', instead, that's 'lived with'.

Several added tests are *not* skipped, that means, when the bug is fixed, they should *not* fire, if they fire, a bug is introduced.

```C#
        [Fact, Trait(Traits.Feature, Traits.Features.Completion)]
        [WorkItem(37361, "https://github.com/dotnet/roslyn/issues/37361")]
        public async Task InTupleNot1()
        {
            var markup = @"
using System;
class C
{
    virtual (int x, C $$
}";

            await VerifyItemIsAbsentAsync(markup, "T");
        }
```

I think we can edit on the issue saying that we added skipped tests, and they should be un-skipped to considered fixed.

Reply:
I would expect the tests to assert expected behavior and be skipped. The only time we've done the "assert the wrong behavior" is if we're trying to assert something doesn't crash as a primary concern. The problem with this:

> Once we fix that, this test will actually fire, demonstrating that yes, we did actually do what we set out to do.

is if somebody else fixes the bug _without realizing it_ they now think they're doing the wrong thing because 'they broke a test'.

Reply:
My main problem has been my continual discovery over the years of skipped tests taht were not unskipped when someone fixed a bug.

It's fairly obvious why this happens.  Very few people check (that includes me btw) for skipped tests when fixing anything.  So they fix goes in and *new* tests are often written, with the old tests sticking around forever (or until someone notices and either unskips or deletes).

Non-running code is not useful code IMO.  

--

On hte other hand, when the tests are *unskipped*, they must be fixed up, and that is a good thing.  IMO, it's a pure win to keep them unskipped.

Reply:
Non-running code isn't useful, but I'd take that any day over a test causing active confusion. :-)

Reply:
Added tuple support.